tree/ntuple/v7/src/RPageStorageFile.cxx

```diff
@@ -575,3 +587,76 @@ ROOT::Experimental::Detail::RPageSourceFile::LoadCluster(DescriptorId_t clusterI
       cluster->SetColumnAvailable(colId);
    return cluster;
 }
+
+
+void ROOT::Experimental::Detail::RPageSourceFile::UnzipClusterImpl(RCluster *cluster)
+{
+   RNTupleAtomicTimer timer(fCounters->fTimeWallUnzip, fCounters->fTimeCpuUnzip);
+   fTaskScheduler->Reset();
+
+   const auto clusterId = cluster->GetId();
+   const auto &clusterDescriptor = fDescriptor.GetClusterDescriptor(clusterId);
+
+   std::vector<std::unique_ptr<RColumnElementBase>> allElements;
+
+   const auto &columnsInCluster = cluster->GetAvailColumns();
+   for (const auto columnId : columnsInCluster) {
+      const auto &columnDesc = fDescriptor.GetColumnDescriptor(columnId);
+
+      allElements.emplace_back(RColumnElementBase::Generate(columnDesc.GetModel().GetType()));
+
+      const auto &pageRange = clusterDescriptor.GetPageRange(columnId);
+      std::uint64_t pageNo = 0;
+      std::uint64_t firstInPage = 0;
+      for (const auto &pi : pageRange.fPageInfos) {
+         ROnDiskPage::Key key(columnId, pageNo);
+         auto onDiskPage = cluster->GetOnDiskPage(key);
+         R__ASSERT(onDiskPage);
+         R__ASSERT(onDiskPage->GetSize() == pi.fLocator.fBytesOnStorage);
+
+         auto taskFunc =
+            [this, columnId, clusterId, firstInPage, onDiskPage,
+             element = allElements.back().get(),
+             nElements = pi.fNElements,
+             indexOffset = clusterDescriptor.GetColumnRange(columnId).fFirstElementIndex
+            ] () {
+               const auto bytesPacked = (element->GetBitsOnStorage() * nElements + 7) / 8;
+               const auto pageSize = element->GetSize() * nElements;
+
+               auto pageBufferPacked = new unsigned char[bytesPacked];
+               if (onDiskPage->GetSize() != bytesPacked) {
+                  fDecompressor(onDiskPage->GetAddress(), onDiskPage->GetSize(), bytesPacked, pageBufferPacked);
+                  fCounters->fSzUnzip.Add(bytesPacked);
+               } else {
+                  // We cannot simply map the onDiskPage because the cluster pool and the page pool have different
+                  // life times
+                  memcpy(pageBufferPacked, onDiskPage->GetAddress(), bytesPacked);
```

Comment:
Ouch. Add a FIXME that we want to hand-over buffers, or does this not work because the pages in the cluster are continuous?